### PR TITLE
do not warn about out-of-date Flutter installation too often

### LINKED ISSUE #9243: "WARNING: your installation of Flutter is old" warning gets old
If someone has decided they don't want to upgrade their Flutter installation (e.g. because we broke their app and they're happy with it as it is), then it gets really annoying to see this every single time you do anything at all with flutter:

``` 
  ╔════════════════════════════════════════════════════════════════════════════╗
  ║ WARNING: your installation of Flutter is 64 days old.                      ║
  ║                                                                            ║
  ║ To update to the latest version, run "flutter upgrade".                    ║
  ╚════════════════════════════════════════════════════════════════════════════╝
```

We should limit the message to once a week or something.

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes #9243

### CODE REVIEW

File: packages/flutter_tools/lib/src/version.dart [hunk 1]
```diff
@@ -244,6 +259,102 @@ class FlutterVersion {
...
+
+    if (versionCheckStamp != null) {
+      // Attempt to parse stamp JSON.
+      try {
+        final dynamic json = JSON.decode(versionCheckStamp);
+        if (json is Map) {
+          printTrace('Warning: expected version stamp to be a Map but found: $json');
+          return fromJson(json);
+        }
+      } catch(error, stackTrace) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/version.dart]: nitty nit nit: missing space after catch

File: packages/flutter_tools/test/src/version_test.dart [hunk 2]
```diff
@@ -25,12 +24,12 @@ final DateTime _upToDateVersion = _testClock.agoBy(FlutterVersion.kVersionAgeCon
 final DateTime _outOfDateVersion = _testClock.agoBy(FlutterVersion.kVersionAgeConsideredUpToDate * 2);
 final DateTime _stampUpToDate = _testClock.agoBy(FlutterVersion.kCheckAgeConsideredUpToDate ~/ 2);
 final DateTime _stampOutOfDate = _testClock.agoBy(FlutterVersion.kCheckAgeConsideredUpToDate * 2);
-const String _stampMissing = '____stamp_missing____';
 
 void main() {
-  group('FlutterVersion', () {
+  group('$FlutterVersion', () {
```
Comment (Hixie) [inline packages/flutter_tools/test/src/version_test.dart]: ?

File: packages/flutter_tools/test/src/version_test.dart [hunk 3]
```diff
@@ -107,26 +137,98 @@ void main() {
...
           errorOnFetch: true,
+          expectServerPing: true,
       );
 
       await version.checkFlutterVersionFreshness();
       _expectVersionMessage('');
     });
   });
+
+  group('$VersionCheckStamp', () {
```
Comment (Hixie) [inline packages/flutter_tools/test/src/version_test.dart]: why the $?

**Comment by Hixie** [comment]
LGTM

@johnmccutchan may wish to review also

File: packages/flutter_tools/lib/src/version.dart (hunk 1 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/version.dart]: Done.

File: packages/flutter_tools/test/src/version_test.dart (hunk 2 above)
Comment (yjbanov) [inline packages/flutter_tools/test/src/version_test.dart]: This is so we get a compiler warning if we rename the class but not the group name.

File: packages/flutter_tools/test/src/version_test.dart (hunk 3 above)
Comment (yjbanov) [inline packages/flutter_tools/test/src/version_test.dart]: Same reason here.

**Pull request merged**